[Metricbeat] Add Linux pressure metricset

b-deam opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

This PR aims to introduce an additional metricset that collects Pressure Stall Information (PSI) for CPU, Memory, and IO.  These are available from kernel >=4.20.

These metrics are incredibly useful for debugging system performance, and more can be read about these metrics below:

- https://www.kernel.org/doc/html/latest/accounting/psi.html
- https://lwn.net/Articles/759658/
- https://facebookmicrosites.github.io/psi/docs/overview.html

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

These are valuable OS metrics that are useful for both debugging and monitoring system performance.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [X] Tests need work.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

- Pull down, run on a Linux host (`ubuntu2004` works from `Vagrantfile`) that:
  - Is on kernel >=4.20  
  - `/proc/pressure` is available on

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes https://github.com/elastic/beats/issues/27223

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-27355/detail/PR-27355/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-27355/detail/PR-27355/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-27355/detail/PR-27355/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-27355/detail/PR-27355/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_27355.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2021-08-13T05:19:23.475Z&rangeTo=2021-08-13T05:39:23.475Z&transactionName=Beats/beats/PR-{number}&transactionType=unknown&latencyAggregationType=avg&traceId=80672a3492af9edba6cb9333a562395b&transactionId=f115a84a107a9042)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2021-08-13T05:29:23.475+0000


* Duration: 14 min 25 sec


* Commit: f3f531a0f14c8eaf1482ad213177dc221b6bc77a




#### Trends :test_tube:


![Image of Build Times](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-27355/buildTimeGraph/png)



</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![4](https://img.shields.io/badge/4%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-27355/detail/PR-27355/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `metricbeat-lint - make -C metricbeat check;make -C metricbeat update;make -C x-pack/metricbeat che`
<ul>
<li>Took 3 min 43 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-27355/runs/1/steps/1205/log/?start=0">here</a></li>
<li>Description: <code>make -C metricbeat check;make -C metricbeat update;make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-lint - make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make -C `
<ul>
<li>Took 2 min 17 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-27355/runs/1/steps/1199/log/?start=0">here</a></li>
<li>Description: <code>make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make -C metricbeat check;make -C metricbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `make check-default`
<ul>
<li>Took 2 min 1 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-27355/runs/1/steps/908/log/?start=0">here</a></li>
<li>Description: <code>make check-default</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-27355/runs/1/steps/1300/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-08-13T05:43:06.270Z] Cleaning up /var/lib/jenkins/workspace/PR-27355-1-fb741e0b-fd61-4025-ac84-64a6950d267f
[2021-08-13T05:43:06.270Z] Change ownership of all files inside the specific folder from root/root to current user/group
[2021-08-13T05:43:06.270Z] ++ id -u
[2021-08-13T05:43:06.270Z] ++ id -g
[2021-08-13T05:43:06.270Z] + docker run -v /var/lib/jenkins/workspace/PR-27355-1-fb741e0b-fd61-4025-ac84-64a6950d267f:/beat alpine:3.4 sh -c 'find /beat -user 0 -exec chown -h 1169:1170 {} \;'
[2021-08-13T05:43:07.212Z] + set -e
[2021-08-13T05:43:07.212Z] + echo 'Change permissions with write access of all files inside the specific folder'
[2021-08-13T05:43:07.212Z] Change permissions with write access of all files inside the specific folder
[2021-08-13T05:43:07.212Z] + chmod -R +w /var/lib/jenkins/workspace/PR-27355-1-fb741e0b-fd61-4025-ac84-64a6950d267f
[2021-08-13T05:43:07.573Z] Running in /var/lib/jenkins/workspace/PR-27355-1-fb741e0b-fd61-4025-ac84-64a6950d267f
[2021-08-13T05:43:10.906Z] Failed in branch x-pack/metricbeat-lint
[2021-08-13T05:43:14.301Z] Error: some files are not up-to-date. Run 'make update' then review and commit the changes. Modified: [metricbeat/docs/fields.asciidoc metricbeat/docs/modules/linux.asciidoc metricbeat/metricbeat.reference.yml metricbeat/module/linux/fields.go metricbeat/module/linux/pressure/pressure.go metricbeat/module/linux/pressure/pressure_test.go]
[2021-08-13T05:43:14.301Z] ../dev-tools/make/mage.mk:21: recipe for target 'check' failed
[2021-08-13T05:43:14.301Z] make: *** [check] Error 1
[2021-08-13T05:43:14.301Z] make: Leaving directory '/var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e/src/github.com/elastic/beats/metricbeat'
[2021-08-13T05:43:14.510Z] Timeout set to expire in 5 min 0 sec
[2021-08-13T05:43:14.841Z] Cleaning up /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e
[2021-08-13T05:43:14.841Z] Change ownership of all files inside the specific folder from root/root to current user/group
[2021-08-13T05:43:14.841Z] ++ id -u
[2021-08-13T05:43:14.841Z] ++ id -g
[2021-08-13T05:43:14.841Z] + docker run -v /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e:/beat alpine:3.4 sh -c 'find /beat -user 0 -exec chown -h 1169:1170 {} \;'
[2021-08-13T05:43:14.841Z] Unable to find image 'alpine:3.4' locally
[2021-08-13T05:43:15.410Z] 3.4: Pulling from library/alpine
[2021-08-13T05:43:15.669Z] c1e54eec4b57: Pulling fs layer
[2021-08-13T05:43:15.928Z] c1e54eec4b57: Verifying Checksum
[2021-08-13T05:43:15.929Z] c1e54eec4b57: Download complete
[2021-08-13T05:43:16.188Z] c1e54eec4b57: Pull complete
[2021-08-13T05:43:16.188Z] Digest: sha256:b733d4a32c4da6a00a84df2ca32791bb03df95400243648d8c539e7b4cce329c
[2021-08-13T05:43:16.188Z] Status: Downloaded newer image for alpine:3.4
[2021-08-13T05:43:17.568Z] + set -e
[2021-08-13T05:43:17.568Z] + echo 'Change permissions with write access of all files inside the specific folder'
[2021-08-13T05:43:17.568Z] Change permissions with write access of all files inside the specific folder
[2021-08-13T05:43:17.568Z] + chmod -R +w /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e
[2021-08-13T05:43:19.048Z] Running in /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e/src/github.com/elastic/beats/build
[2021-08-13T05:43:19.369Z] + rm -rf ve
[2021-08-13T05:43:19.369Z] + find . -type d -name vendor -exec rm -r {} ;
[2021-08-13T05:43:19.745Z] + python .ci/scripts/pre_archive_test.py
[2021-08-13T05:43:19.745Z] Copy ./metricbeat/build into build/metricbeat/build
[2021-08-13T05:43:19.772Z] Running in /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e/src/github.com/elastic/beats/build
[2021-08-13T05:43:19.822Z] Recording test results
[2021-08-13T05:43:20.439Z] None of the test reports contained any result
[2021-08-13T05:43:20.440Z] [Checks API] No suitable checks publisher found.
[2021-08-13T05:43:20.792Z] + tar --version
[2021-08-13T05:43:21.171Z] + tar --exclude=test-build-artifacts-metricbeat-lint-tgz -czf test-build-artifacts-metricbeat-lint-tgz .
[2021-08-13T05:43:21.608Z] + gsutil --version
[2021-08-13T05:43:24.970Z] Masking supported pattern matches of $FILE_CREDENTIAL
[2021-08-13T05:43:25.468Z] + gcloud auth activate-service-account --key-file ****
[2021-08-13T05:43:26.407Z] Activated service account credentials for: [beats-ci-gcs-plugin@elastic-ci-prod.iam.gserviceaccount.com]
[2021-08-13T05:43:26.787Z] + gsutil -m -q cp -a public-read test-build-artifacts-metricbeat-lint-tgz gs://beats-ci-temp/Beats/beats/PR-27355-1
[2021-08-13T05:43:29.227Z] + python .ci/scripts/search_system_tests.py
[2021-08-13T05:43:29.319Z] [INFO] system-tests=''. If no empty then let's create a tarball
[2021-08-13T05:43:29.790Z] + go clean -modcache
[2021-08-13T05:43:33.125Z] Timeout set to expire in 5 min 0 sec
[2021-08-13T05:43:33.455Z] Cleaning up /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e
[2021-08-13T05:43:33.455Z] Change ownership of all files inside the specific folder from root/root to current user/group
[2021-08-13T05:43:33.455Z] ++ id -u
[2021-08-13T05:43:33.455Z] ++ id -g
[2021-08-13T05:43:33.455Z] + docker run -v /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e:/beat alpine:3.4 sh -c 'find /beat -user 0 -exec chown -h 1169:1170 {} \;'
[2021-08-13T05:43:34.394Z] + set -e
[2021-08-13T05:43:34.394Z] + echo 'Change permissions with write access of all files inside the specific folder'
[2021-08-13T05:43:34.394Z] Change permissions with write access of all files inside the specific folder
[2021-08-13T05:43:34.394Z] + chmod -R +w /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e
[2021-08-13T05:43:34.751Z] Running in /var/lib/jenkins/workspace/PR-27355-1-5e024af8-265d-41e2-a81b-91c1a9dc8a0e
[2021-08-13T05:43:37.758Z] Failed in branch metricbeat-lint
[2021-08-13T05:43:38.234Z] + go clean -modcache
[2021-08-13T05:43:40.814Z] Timeout set to expire in 5 min 0 sec
[2021-08-13T05:43:41.146Z] Cleaning up /var/lib/jenkins/workspace/Beats_beats_PR-27355
[2021-08-13T05:43:41.146Z] Change ownership of all files inside the specific folder from root/root to current user/group
[2021-08-13T05:43:41.146Z] ++ id -u
[2021-08-13T05:43:41.146Z] ++ id -g
[2021-08-13T05:43:41.146Z] + docker run -v /var/lib/jenkins/workspace/Beats_beats_PR-27355:/beat alpine:3.4 sh -c 'find /beat -user 0 -exec chown -h 1169:1170 {} \;'
[2021-08-13T05:43:41.146Z] Unable to find image 'alpine:3.4' locally
[2021-08-13T05:43:41.713Z] 3.4: Pulling from library/alpine
[2021-08-13T05:43:41.972Z] c1e54eec4b57: Pulling fs layer
[2021-08-13T05:43:42.231Z] c1e54eec4b57: Verifying Checksum
[2021-08-13T05:43:42.231Z] c1e54eec4b57: Download complete
[2021-08-13T05:43:42.490Z] c1e54eec4b57: Pull complete
[2021-08-13T05:43:42.490Z] Digest: sha256:b733d4a32c4da6a00a84df2ca32791bb03df95400243648d8c539e7b4cce329c
[2021-08-13T05:43:42.490Z] Status: Downloaded newer image for alpine:3.4
[2021-08-13T05:43:43.428Z] + set -e
[2021-08-13T05:43:43.428Z] + echo 'Change permissions with write access of all files inside the specific folder'
[2021-08-13T05:43:43.428Z] Change permissions with write access of all files inside the specific folder
[2021-08-13T05:43:43.428Z] + chmod -R +w /var/lib/jenkins/workspace/Beats_beats_PR-27355
[2021-08-13T05:43:43.555Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-27355
[2021-08-13T05:43:46.314Z] Stage "Build&Test" skipped due to earlier failure(s)
[2021-08-13T05:43:46.419Z] Stage "Extended" skipped due to earlier failure(s)
[2021-08-13T05:43:46.528Z] Stage "Packaging" skipped due to earlier failure(s)
[2021-08-13T05:43:46.643Z] Stage "Packaging-Pipeline" skipped due to earlier failure(s)
[2021-08-13T05:43:46.808Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-27355/src/github.com/elastic/beats
[2021-08-13T05:43:47.524Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-27355
[2021-08-13T05:43:47.646Z] [INFO] getVaultSecret: Getting secrets
[2021-08-13T05:43:47.699Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-08-13T05:43:48.824Z] + chmod 755 generate-build-data.sh
[2021-08-13T05:43:48.824Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-27355/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-27355/runs/1 FAILURE 865071
[2021-08-13T05:43:48.824Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-27355/runs/1/steps/?limit=10000 -o steps-info.json
[2021-08-13T05:43:49.374Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-27355/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-08-13T05:43:49.375Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-08-13T05:43:50.718Z] Retry 2/3 exited 22, retrying in 2 seconds...
[2021-08-13T05:43:52.579Z] Retry 3/3 exited 22, no more retries left.
[2021-08-13T05:43:52.579Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-27355/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



## :grey_exclamation: Flaky test report
No test was executed to be analysed.



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

fearful-symmetry commented on metricbeat/module/linux/pressure/_meta/fields.yml:
```diff
@@ -0,0 +1,90 @@
+      type: group
+      description: Linux pressure stall information
+      fields:
+        - name: cpu.some.10
```
In `system/process` cgroup data, we're reporting these as `10.pct` and so on. We'll also need to add `format: percent`

b-deam replied:
Cool. I tried to follow similar conventions for `total`, in that it becomes `total.time.us` - does this make sense here?

fearful-symmetry commented on metricbeat/module/linux/pressure/pressure.go:
```diff
@@ -0,0 +1,122 @@
+		return nil, fmt.Errorf("the %v/%v metricset is only supported on Linux", moduleName, metricsetName)
+	}
+
+	fs, err := procfs.NewFS("/proc")
```
We have a global `Path.Hostfs` that's used in other places, and we'll probably want to use it here, as it allows users to specify a custom root, for instances where they're running MB inside a container.

b-deam replied:
Yup! I've copied the implementation from the other modules.

fearful-symmetry commented on metricbeat/module/linux/pressure/pressure.go:
```diff
@@ -0,0 +1,122 @@
+	return nil
+}
+
+func FetchLinuxPSIStats(m *MetricSet) ([]common.MapStr, error) {
```
I don't think this needs to be exported?

b-deam replied:
You're right! Fixed.

fearful-symmetry commented on metricbeat/module/linux/pressure/pressure_test.go:
```diff
@@ -0,0 +1,55 @@
+	mbtest "github.com/elastic/beats/v7/metricbeat/mb/testing"
+)
+
+func TestFetch(t *testing.T) {
```
A mocked procfs root will also be useful here, as we can just specify a testdata file.

b-deam replied:
Great idea. Thanks for the tip!

fearful-symmetry reviewed (commented):
This is looking good! Just a few small things, mostly beats cruft, we need to change. If you want, I can take the PR over from here; I don't want to make you feel like you're obligated to spend your time coding to the various metricbeat idioms.

b-deam commented:
>  If you want, I can take the PR over from here; I don't want to make you feel like you're obligated to spend your time coding to the various metricbeat idioms.

Not a problem - I _think_ I've addressed the comments so far but feel free to make any small changes if I've otherwise messed things up there.

fearful-symmetry commented:
Real quick, you'll need to run `mage fmt update` in the main metricbeat directory.

b-deam commented:
> Real quick, you'll need to run `mage fmt update` in the main metricbeat directory.

Hmm.. I tried this but I don't see any changes. Any ideas?

```
bradleydeam in ~/go/src/github.com/b-deam/beats/metricbeat on branch linux-pressure-metricset $ make update  
mage update
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/fields.yml
>> Building metricbeat.yml for linux/amd64
>> Building metricbeat.reference.yml for linux/amd64
>> Building metricbeat.docker.yml for linux/amd64
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/build/fields/fields.all.yml
exec: go list -m
exec: go list -m              

bradleydeam in ~/go/src/github.com/b-deam/beats/metricbeat on branch linux-pressure-metricset $ mage fmt update
>> fmt - go-licenser: Adding missing headers
>> fmt - goimports: Formatting Go code
>> fmt - autopep8: Formatting Python code
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/fields.yml
>> Building metricbeat.yml for linux/amd64
>> Building metricbeat.reference.yml for linux/amd64
>> Building metricbeat.docker.yml for linux/amd64
Generated fields.yml for metricbeat to /Users/bradleydeam/go/src/github.com/b-deam/beats/metricbeat/build/fields/fields.all.yml
exec: go list -m
exec: go list -m

bradleydeam in ~/go/src/github.com/b-deam/beats/metricbeat on branch linux-pressure-metricset $ git status
On branch linux-pressure-metricset
Your branch is up to date with 'origin/linux-pressure-metricset'.

nothing to commit, working tree clean
```

fearful-symmetry commented:
@b-deam I think you'll need to re-merge upstream changes? A few things may have changed in the doc build.

b-deam commented:
> @b-deam I think you'll need to re-merge upstream changes? A few things may have changed in the doc build.

🤦 - ofc. Done. 🤞 tests are good!

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

fearful-symmetry commented on metricbeat/module/linux/pressure/pressure.go:
```diff
@@ -0,0 +1,143 @@
+		return nil, errors.New("unexpected module type")
+	}
+
+	path := filepath.Join(linuxModule.HostFS, "proc")
```
I...didn't even know that was there. There's actually a `"github.com/elastic/beats/v7/libbeat/paths"` that's global and used by most things, but I guess linux has its own hostfs path. In the future we might want to change it, but for now we should probably keep linux internally consistent.

fearful-symmetry commented on metricbeat/module/linux/pressure/pressure_test.go:
```diff
@@ -0,0 +1,95 @@
+
+	resources := []string{"cpu", "memory", "io"}
+
+	for i, _ := range events {
```
Nit: This can be simplified to `for i := range events`

fearful-symmetry reviewed (approved):
LGTM.

b-deam commented:
> LGTM.

I re-ran a test to make sure it all looked good and noticed that I'd messed up the fields/mapping so that the metrics were actually recorded as:

```
linux.pressure.pressure.*
```

This should be fixed now..

b-deam commented:
Thanks for all the help @fearful-symmetry!
